### PR TITLE
Spin v2.1.0 Documentation

### PR DESCRIPTION
**Pull request by tpmccallum**
Content must go through a pre-merge checklist.

## Pre-Merge Content Checklist

This documentation has been checked to ensure that:

- [x] The `title, `template`, and `date` are all set
- [x] Does this PR have a new menu item (anywhere in `templates/*.hbs` files) that points to a document `.md` that is set to publish in the future? If so please only publish the `.md` and `.hbs` changes in real-time (otherwise there will be a menu item pointing to a `.md` file that does not exist)
- [x] File does not use CRLF, but uses plain LF (hint: use `cat -ve <filename> | grep '^M' | wc -l` and expect 0 as a result) 
- [x] Has passed [`bart check`](https://developer.fermyon.com/bartholomew/quickstart)
- [x] Has been manually tested by running in Spin/Bartholomew (hint: use `PREVIEW_MODE=1` and run `npm run styles` to update styling)
- [x] Headings are using Title Case
- [x] Code blocks have the programming language set to properly highlight syntax and the proper copy directive
- [x] Have tested with `npm run test` and resolved all errors
- [x] Relates to an existing (potentially outdated) blog article? If so please add URL in blog to point to this content.

### CODE REVIEW

File: content/spin/v2/kv-store-api-guide.md [hunk 1]
```diff
@@ -57,7 +57,7 @@ fn handle_request(_req: Request) -> Result<impl IntoResponse> {
     store.set("mykey", "myvalue")?;
     let value = store.get("mykey")?;
     let response = value.unwrap_or_else(|| "not found".into());
-    Ok(http::Response::builder().status(200).body(response)?)
+    Ok(Response::builder().status(200).body(response)?)
```
Comment (rylev) [inline content/spin/v2/kv-store-api-guide.md:60]: This is not correct. The sdk `Response` builder does not return a `Result` from the `body` method like the `http` crate's `ResponseBuilder` does. 

This should be the following I believe (but I didn't run this through a compiler so you should probably do that to make sure this works):

```rust
Ok(Response::builder().status(200).body(response).build())
```

**Review by rylev — CHANGES_REQUESTED** [review]
It seems you removed the http crate assuming that the API is the same as the `spin_sdk::http::ResponseBuilder`, but it's not.

**Review by itowlson — CHANGES_REQUESTED** [review]

File: content/spin/v2/kv-store-api-guide.md (hunk 1 above)
Comment (tpmccallum) [reply content/spin/v2/kv-store-api-guide.md:60]: Thanks,
I will update these and test theem also.
Great spotting; thank you!!!

File: content/spin/v2/redis-trigger.md [hunk 2]
```diff
@@ -30,7 +30,7 @@ component = "my-application"  # the name of the component to handle this route
 
 Such a trigger says that Redis messages on the specified _channel_ should be handled by the specified _component_. The `component` field works the same way across all triggers - see [Triggers](triggers) for the details.
 
-> Spin subscribes only to the channels that are mapped to components. Other channels are ignored.
+> Spin subscribes only to the channels that are mapped to components. Other channels are ignored. Individual Redis channels can be subscribed to by one or more components (A 1:N relationship exists between an individual component and Redis channels).
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/redis-trigger.md:33]: Thanks @itowlson 
I have used this rewording in the next push.

File: content/spin/v2/rust-components.md [hunk 3]
```diff
@@ -289,6 +291,50 @@ proxies or URL shorteners.
 
 > The Spin SDK for Rust provides more flexibility than we show here, including allowing streaming uploads or downloads. See the [Outbound HTTP API Guide](./http-outbound.md) for more information.
 
+## Routing in a Component
+
+The Rust SDK [provides a router](https://github.com/fermyon/spin/tree/main/examples/http-rust-router) that makes it easier to handle synchronous and asynchronous routing within a component:
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/rust-components.md:296]: Thanks, @itowlson 
Have updated this text in the next push.

[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/rust-components.md:296]: Shall do, as part of next push.
Thanks!

File: content/spin/v2/redis-trigger.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/redis-trigger.md:33]: New changes coming in next push.

File: content/spin/v2/rust-components.md [hunk 4]
```diff
@@ -289,6 +291,50 @@ proxies or URL shorteners.
...
+mod api {
+    use super::*;
+
+    // /goodbye/:planet
+    pub fn goodbye_planet(_req: Request, params: Params) -> Result<impl IntoResponse> {
+        let planet = params.get("planet").expect("PLANET");
+        Ok(Response::new(200, planet.to_string()))
+    }
+
+    // /hello/:planet
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/rust-components.md:324]: Fair point, removed it for now.
Thanks @itowlson

File: content/spin/v2/serverless-ai-hello-world.md [hunk 5]
```diff
@@ -224,13 +224,13 @@ Now let's use the Spin SDK to access the model from our app. Executing inference
...
+use spin_sdk::{http::{IntoResponse, Request, Response}, http_component, llm};
 
 #[http_component]
 fn hello_world(_req: Request) -> anyhow::Result<impl IntoResponse> {
    let model = llm::InferencingModel::Llama2Chat;
    let inference = llm::infer(model, "Can you tell me a joke about cats");
-   Ok(http::Response::builder()
+   Ok(Response::builder()
        .status(200)
        .body(format!("{:?}", inference))?)
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/serverless-ai-hello-world.md:235]: Yes, this is updated in the next push.
Thanks!
BTW I am testing all of these thoroughly by creating new apps.

File: content/spin/v2/cli-reference.md [hunk 6]
```diff
@@ -710,6 +1161,34 @@ OPTIONS:
...
+$ spin registry push --help 
+
+spin-registry-push 
+Push a Spin application to a registry
+
+USAGE:
+    spin registry push [OPTIONS] <REFERENCE>
+
+ARGS:
+    <REFERENCE>    Reference of the Spin application
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/cli-reference.md:1178]: Thanks @itowlson - new changes in next push.

File: content/spin/v2/cli-reference.md [hunk 7]
```diff
@@ -675,6 +1103,29 @@ OPTIONS:
...
+$ spin registry pull --help
+
+spin-registry-pull 
+Pull a Spin application from a registry
+
+USAGE:
+    spin registry pull [OPTIONS] <REFERENCE>
+
+ARGS:
+    <REFERENCE>    Reference of the Spin application
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/cli-reference.md:1120]: Thanks @itowlson - new changes in next push

File: content/spin/v2/cli-reference.md [hunk 8]
```diff
@@ -645,6 +1048,31 @@ OPTIONS:
...
+$ spin registry login --help
+
+spin-registry-login 
+Log in to a registry
+
+USAGE:
+    spin registry login [OPTIONS] <SERVER>
+
+ARGS:
+    <SERVER>
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/v2/cli-reference.md:1065]: Thanks @itowlson - changes in next push.

File: content/spin/v2/quickstart.md [hunk 9]
```diff
@@ -325,17 +325,18 @@ annotated with the `http_component` macro which identifies it as the entry point
...
 
 ```rust
-use spin_sdk::http::{IntoResponse, Request};
+use spin_sdk::http::{IntoResponse, Request, Response};
 use spin_sdk::http_component;
 
 /// A simple Spin HTTP component.
 #[http_component]
-fn handle_hello_rust(req: Request) -> anyhow::Result<impl IntoResponse> {
+fn handle_quickstart(req: Request) -> anyhow::Result<impl IntoResponse> {
```
Comment (itowlson) [inline content/spin/v2/quickstart.md:333]: The generated code will be `handle_hello_rust`.

File: content/spin/v2/writing-apps.md [hunk 10]
```diff
@@ -139,15 +139,22 @@ Multiple components can have the same source.  An example is a document archive,
 
 At the Wasm level, a Spin component is a Wasm component or module that exports a handler for the application trigger.  At the developer level, a Spin component is a library or program that implements your event handling logic, and uses Spin interfaces, libraries, or tools to associate that with the events handled by Spin.
 
-See the Language Guides section for how to do this in your preferred language. As an example, this is a component written in the Rust language. The `hello_world` function uses an attribute `#[http_component]` to identify the function as handling a Spin HTTP event. The function takes a `Request` and returns a `Result<Response>`:
+See the Language Guides section for how to do this in your preferred language. As an example, this is a component written in the Rust language. The `hello_world` function uses an attribute `#[http_component]`.
```
Comment (itowlson) [inline content/spin/v2/writing-apps.md:142]: Not sure why the explanation of the attribute and signature were removed here; without them this paragraph seems a bit bare.

**Review by itowlson — APPROVED** [review]
Looks good except for a couple of minor things - thanks!

File: content/spin/v2/writing-apps.md (hunk 10 above)
Comment (tpmccallum) [inline content/spin/v2/writing-apps.md:142]:
```suggestion
See the Language Guides section for how to do this in your preferred language. As an example, this is a component written in the Rust language. The `hello_world` function uses an attribute `#[http_component]` to identify the function as handling a Spin HTTP event. The function takes a `Request` and returns a `anyhow::Result<impl IntoResponse>`.
```

File: content/spin/v2/quickstart.md (hunk 9 above)
Comment (tpmccallum) [inline content/spin/v2/quickstart.md:333]:
```suggestion
fn handle_hello_rust(req: Request) -> anyhow::Result<impl IntoResponse> {
```

File: content/spin/v2/rust-components.md [hunk 11]
```diff
@@ -298,9 +339,9 @@ components.
 Let's see how we can use the Rust SDK to connect to Redis:
 
 ```rust
-use anyhow::{Context, Result};
+use anyhow::{Context};
```
Comment (rylev) [inline content/spin/v2/rust-components.md:342]: Nit: the syntax used here is allowed, but `{}` are not normally used when only importing one thing from a module. Instead: 

```suggestion
use anyhow::Context;
```

**Review by rylev — APPROVED** [review]

**Review by dicej — APPROVED** [review]

**Pull request merged**